src/iris/api.py

```diff
@@ -2695,6 +2697,69 @@ def on_get(self, req, resp, app_name):
         resp.body = ujson.dumps({'key': key})
 
 
+class ApplicationSecondaryKey(ApplicationKey):
+    allow_read_no_auth = False
+
+    def on_get(self, req, resp, app_name):
+        if not req.context['username']:
+            raise HTTPUnauthorized('You must be a logged in user to view this app\'s key')
+
+        with db.guarded_session() as session:
+            if not req.context['is_admin']:
+                has_permission = session.execute(
+                    '''SELECT 1
+                       FROM `application_owner`
+                       JOIN `target` on `target`.`id` = `application_owner`.`user_id`
+                       JOIN `application` on `application`.`id` = `application_owner`.`application_id`
+                       WHERE `target`.`name` = :username
+                       AND `application`.`name` = :app_name''',
+                    {'app_name': app_name, 'username': req.context['username']}).scalar()
+                if not has_permission:
+                    raise HTTPForbidden('You don\'t have permissions to view this app\'s key.')
+
+            key = session.execute(
+                'SELECT `secondary_key` FROM `application` WHERE `name` = :app_name LIMIT 1',
+                {'app_name': app_name}).scalar()
+            session.close()
+
+        resp.body = ujson.dumps({'key': key})
+
+    def on_post(self, req, resp, app_name):
+        # Only admins and application owners can generate a secondary key
+        if not req.context['username']:
+            raise HTTPUnauthorized('You must be a logged in user to app')
+
+        with db.guarded_session() as session:
+            if not req.context['is_admin']:
+                has_permission = session.execute(
+                    '''SELECT 1
+                       FROM `application_owner`
+                       JOIN `target` on `target`.`id` = `application_owner`.`user_id`
+                       JOIN `application` on `application`.`id` = `application_owner`.`application_id`
+                       WHERE `target`.`name` = :username
+                       AND `application`.`name` = :app_name''',
+                    {'app_name': app_name, 'username': req.context['username']}).scalar()
+                if not has_permission:
+                    raise HTTPForbidden('You don\'t have permissions to re-key this app.')
+
+        data = {
+            'app_name': app_name,
+            'new_key': hashlib.sha256(os.urandom(32)).hexdigest()
```

Comment:
As we're now doing this twice, we should probably abstract the key generation into a separate function.